### PR TITLE
Molecular Mechanics Induction Energy Calculator 

### PR DESCRIPTION
**Pull request by shehan807**
## Description
**The goal of this program is to simplify the calculation of the induction energy component using a SAPT-based, ab initio polarizable molecular dynamics force field function form.** The theoretical formalism is elaborated in [Shehan's U_pol Repo](https://github.com/shehan807/U_pol). In essence, each relevant atom, considered a "core" atom, is paired with a "shell" or Drude particle through which induced-dipole/induced-dipole interactions manifest. The added functions have been verified to reproduce polarization energies in [OpenMM](https://github.com/openmm/openmm) with the advantage of a Drude SCF optimization scheme implemented using [JAX](https://github.com/jax-ml/jax). 

## Summary
  **Plugin (`crystalatte/plugins/`)**
  - `polarization_ff.py` — the Drude FF induction plugin. All the JAX/Drude stuff lives here: the SCF optimizer (`drudeOpt`), the induction kernel (`Uind`), and the CrystaLattE entry point `polarization_energy_function` (writes `nmer["nambe"]` in Hartree via inclusion–exclusion).
  - `utils.py` — FF I/O + graph helpers (`XmlMD`, `get_pol_params`, isomorphism stuff). 
  - `force_fields.py` — trimmed back to just the lightweight`example_energy_function` template (prints gated on `verbose`).
  - `__init__.py` — each plugin imported under its own guarded `try/except`, so a missing dep only disables that one plugin.

  **Example (`examples/polarization_ff/`)**
  - `polarization_ff_imidazole.py` — runs imidazole CLE through the plugin, up to 3-body.
  - `imidazole/` — inputs (`.pdb`, `.xml`, `_map.csv`).
  - `pol_ff_cle.yml` — portable conda env, replaces `p4_jax_mda.yml`.

  **Test**
  - `test_96_polarization_ff.py` *(new)* — 2-body regression through `crystalatte.main()`'s custom dispatch. Checks dimer count, replicas, per-dimer `nambe`, and total CLE. `importorskip`-guarded, cleans up after itself.

  **Data/config**
  - `Imidazole.cif` — input structure.
  - `.gitignore` — ignores run artifacts.

## Status
- [x] Ready to go

### CODE REVIEW

File: crystalatte/plugins/force_fields.py [hunk 1]
```diff
@@ -1,6 +1,468 @@
+from . import utils
+from optax import safe_norm
+from jaxopt import NonlinearCG
+import jax.numpy as jnp
+import jax
+from jax import jit
```
Comment (loriab) [inline crystalatte/plugins/force_fields.py]: There's a lot of extra dependencies with this, so I think it should go into its own file

File: crystalatte/plugins/__init__.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 try:
+    from . import openmm_utils
```
Comment (loriab) [inline crystalatte/plugins/__init__.py]: You'll want these as separate blocks. try/import-A/except/pass  try/import-B/except/pass so the attempts are independent. What's openmm_utils — I don't see it here?

File: crystalatte/plugins/force_fields.py [hunk 3]
```diff
@@ -1,6 +1,468 @@
...
+                  + Qi_shell * Qj_shell / |R_ij + D_i|
+                  + Qi_shell * Qj_shell / |R_ij + D_j|
+                  + Qi_shell * Qj_shell / |R_ij + D_i - D_j|
+                  ]
+
+    where:
+    - 1/4πε₀ is the Coulomb constant
+    - Qi_shell, Qj_shell are the Drude charges on shared molecules 
+    - R_ij is the distance matrix between sites i and j
+    - D_i, D_j are the displacements of Drude particles from parent atom nuclei
```
Comment (loriab) [inline crystalatte/plugins/force_fields.py]: thanks for the nice docstrings

File: examples/polarization_ff/imidazole/imidazole_outputs/s4machinefile-79897 [hunk 4]
```diff
@@ -0,0 +1,6 @@
+ds1
```
Comment (loriab) [inline examples/polarization_ff/imidazole/imidazole_outputs/s4machinefile-79897]: there's a lot of files added in this PR that maybe just got caught by a generic `git add`. I'd go through and remove any of the molecule or output files that aren't going to be used in a test or have a definite purpose. s4machinefiles can definitely go :-)

**Review by loriab — COMMENTED** [review]
Thanks, looks like an interesting approach.

File: crystalatte/plugins/__init__.py (hunk 2 above)
Comment (shehan807) [reply crystalatte/plugins/__init__.py]: openmm_utils was removed, and now this line is deleted.

File: crystalatte/plugins/force_fields.py (hunk 1 above)
Comment (shehan807) [reply crystalatte/plugins/force_fields.py]: Done — moved all the JAX/Drude polarization code into its own module, `crystalatte/plugins/polarization_ff.py

**Comment by shehan807** [comment]
Hi @loriab, thanks for your patience on this! I've addressed all your comments, trimmed the PR down to just the essential functions under `crystalatte/plugins/`, and isolated the heavy JAX/Drude deps into their own module (`polarization_ff.py`) behind guarded imports so they don't load with the base package. Added a regression test and a runnable example too.

CC'ing @Awallace3 — you might want to review `polarization_energy_function` wiring, which lets stock crystalatte dispatch to a custom energy function like the polarization FF.
